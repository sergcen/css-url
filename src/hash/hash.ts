import * as crypto from 'crypto';
import * as xxh from 'xxhashjs';

const HEXBASE = 16;

export type THashAlg = 'xxhash32' | 'xxhash64' | string | undefined;

const getxxhash = (content: Buffer | string, method: THashAlg = 'xxhash32') => {
    const hashFunc = method === 'xxhash32' ? xxh.h32 : xxh.h64;
    const seed = 0;

    return hashFunc(seed).update(content).digest().toString(HEXBASE);
};

export const hash: (content: Buffer | string, alg?: THashAlg, shrink?: number | undefined) => string = (
    content,
    method = 'xxhash32',
    shrink = 8,
) => {
    let result;
    if (method === 'xxhash32' || method === 'xxhash64') {
        result = getxxhash(content, method);
    } else {
        const hashFunc = crypto.createHash(method || 'sha256');

        result = hashFunc.update(content).digest('hex');
    }

    return result.slice(0, shrink);
};
