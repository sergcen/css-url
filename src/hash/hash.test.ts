import { hash } from './hash';

describe('hash', () => {
    it.only('base', () => {
        expect(hash('some content str')).toMatchSnapshot();
        expect(hash('some content str', 'xxhash32', 10)).toMatchSnapshot();
        expect(hash('some content str', 'xxhash64', 100)).toMatchSnapshot();
        expect(hash('some content str', 'sha512', 20)).toMatchSnapshot();
        expect(hash(Buffer.from('some content str'))).toMatchSnapshot();
    });
})
