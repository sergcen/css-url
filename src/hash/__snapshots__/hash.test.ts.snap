// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash base 1`] = `"586a786b"`;

exports[`hash base 2`] = `"586a786b"`;

exports[`hash base 3`] = `"b7d90b16ad9d9aab"`;

exports[`hash base 4`] = `"a16ed3efd431abe32575"`;

exports[`hash base 5`] = `"586a786b"`;
